### DISCUSSION TITLE
OpenRarity beta release

### DISCUSSION THREAD

**Discussion by impreso** [OpenRarity/open-rarity#30] [Announcements]
Public beta release of OpenRarity library discussion.

**Comment by alexandresobolevski** [comment]
Hi! I built the rarity algos at Zapper and would like to help out in any way possible.

**Reply by impreso** [reply]
Hey, @alexandresobolevski thanks for reaching out - we are refining our roadmap and the governance process for external developers. Stay tuned for the updates!

**Reply by impreso** [reply]
@alexandresobolevski if you have any feedback to share on the rarity algorithm we would be interested in talking to you as well.

**Reply by alexandresobolevski** [reply]
The only thing that I have to say is that, while I understand completely the reasoning why "meta traits" such as trait counts are not considered in this initiative; as an NFT collector I definitely consider it in my rarity assessment. And we'd need the rarity library to support what collectors/explorers want to consider. A "clean" nft with few traits and the only one with a very low number of traits is commonly used as an argument to value an NFT. So objectively speaking, as much as I want to switch to an open sourced rarity library, I'd have to think about it as I'd lose some value there. And I am not fully convinced by the counter argument of expecting NFT creators to create/update metadata with the idea that anything that should be considered for rarity should be there.

The algorithm itself makes sense, we're doing the same thing: normalization for trait cardinality, etc. and I see the [probability to have a null trait value is also considered](https://github.com/OpenRarity/open-rarity/blob/a643f520a4c5e3e3593faf9e64a73aaccb12cdce/open_rarity/scoring/utils.py#L40-L42), so I am happy to see that everything is aligned there.

**Reply by impreso** [reply]
@alexandresobolevski we added the trait count recently to the lib.

**Comment by jakenft** [comment]
Hi there! This looks great! Where is the metadata coming from in this system? We at NFT Protocol (nft.org) operate an enterprise-grade on- and off-chain indexer that maintains a GraphQL database of metadata for 100% of tokens that exist on Ethereum and Polygon. The indexer extracts metadata from every new block and also completes a rescan of the entirety of both chains for metadata updates every 48 hours. If this could improve collection metadata sourcing for OpenRarity I'd love to further discuss this! jake@nft.org

**Reply by impreso** [reply]
`Where is the metadata coming from in this system` -> consumers of the library will provide the metadata and invoke ranker library API. We built a helper function that resolves metadata from OpenSea API as well.
`If this could improve collection metadata sourcing for OpenRarity I'd love to further discuss this` -> if you have an api , you could propose a PR with the helper tool that resolves metadata from the nft.org APIs.

**Comment by serge555** [comment]
Hey! We are https://imxflow.com - nft marketplace & rarity tool for Layer 2 NFT protocol Immutable X. Happy to be part of the community.
@imxflowcom

**Reply by impreso** [reply]
Thanks for reaching out. Let us know if you have any feedback on algorithm or library implementation. This would be the best way to start the collaboration! File feature requests or bugs in github if you find any issues.

**Reply by serge555** [reply]
Sorry, a lot of work right now, I'll be back with a feedback soon. It would be great if there was a document with an algorithm and formula. Watched the video, but haven't figured it out yet.

**Reply by impreso** [reply]
Did you look at our developer docs? https://openrarity.gitbook.io/developers

**Reply by 365sma11** [reply]
> Did you look at our developer docs? https://openrarity.gitbook.io/developers

An example with the [formula](https://openrarity.gitbook.io/developers/fundamentals/methodology) would be helpful. thx

**Reply by 365sma11** [reply]
> Did you look at our developer docs? https://openrarity.gitbook.io/developers

Also another thing that would be helpful on your [Methodology page](https://openrarity.gitbook.io/developers/fundamentals/methodology), a comparison of OpenRarity Rarity score for an NFT in comparison to Other common rarity calculated Rarity Scores.  One to include is a different rarity scoring calc I have seen which would be interesting to compare is [NFTGo ](https://www.publish0x.com/navid-ladani/explaining-nft-rarity-score-models-xmmorpr#:~:text=Counting%20the%20average%20value%20of%20the%20Jaccard%20Distance%2C,get%20the%20final%20NFTGO%27s%20rarity%20score%20of%20BAYC%231154.)

**Comment by jeevanpillay0000** [comment]
Hey all, congrats on the launch! Amazing stuff.

Just a quick q, wondering if there are any talks about a js/ts version of this.

**Reply by impreso** [reply]
@jeevanpillay  we are working on it. Please follow https://github.com/OpenRarity/open-rarity/issues/53

**Reply by jeevanpillay0000** [reply]
awesome! thanks mate.

**Comment by poisonouspinups** [comment]
We would love to be involved 
Great community
Growimg Universe of #NFT projects 
Great rewards
https://poisonous-cherry-bombs-mint.vercel.app/

**Comment by grigb** [comment]
Interested in cross-chain support for emerging chains.

**Reply by impreso** [reply]
We do have cross-chain support.

**Comment by thedanielmark** [comment]
Hi! I run @nftconomy - an NFT data analytics platform and we're interested in becoming a launch partner after integrating with OpenRarity.

**Reply by impreso** [reply]
Nice to hear  - you can integrate it in your app. OpenRarity is free and distributed under Apache 2.0 license model.

**Reply by impreso** [reply]
https://github.com/OpenRarity/open-rarity/tree/main/icons/OpenRarity_Icons - marketing materials. https://openrarity.gitbook.io/developers/quick-guides/integrating-openrarity-in-your-application - application integration guide.

**Comment by anbergo** [comment]
Hi! this is @BendDAO - BendDAO is the first NFT liquidity protocol supporting instant NFT-backed loans, Collateral Listing, and NFT Down Payment. We're interested in becoming a launch partner.

**Reply by impreso** [reply]
Nice! How exactly you plan to support OpenRarity in your stack? All marketing resources are here - https://github.com/OpenRarity/open-rarity/tree/main/icons/OpenRarity_Icons .  https://openrarity.gitbook.io/developers/quick-guides/integrating-openrarity-in-your-application - application integration guide. Let me know if this is enough info!

**Comment by nmelentjevs** [comment]
Hi, cnftjungle.io here, would like to offer our help as well in any way possible! We do have all Cardano NFT collections in our db if you need any test data to run against.

**Reply by impreso** [reply]
Hey @nmelentjevs, We built the library in a multi-chain way , so you can use it as long as Cardano NFTs adhere to the the metadata standard.

**Reply by impreso** [reply]
happy to take any contributions to enable OR on Cardano 👍

**Reply by nmelentjevs** [reply]
Awesome, will surely take a look at the docs, appreciate it!

**Comment by erzich** [comment]
Hello - Thanks for all you are doing...  I appreciate your hard work. 

It seems that the math is wrong on at least one of the collections for which you are ranking rarity.  There is some discrepancy and it causes confusion...  it seems pretty clear, and I wonder if you would be open to double checking it...  I'm not sure if this is the right place to reach out, but I do believe you are interested in accuracy so I thought you might be interested.

Thanks in advance.  I look forward to hearing back.

**Reply by impreso** [reply]
Thanks for reaching out  @erzich , mind sharing collection name?

**Reply by erzich** [reply]
Sure.  Thanks for getting back to me.

the collection is slycrocs.

https://opensea.io/collection/slycrocs

When you just look at the traits, there are a couple of clear differentiators for rarity, where the traits are below .5% of the collection.  One trait in particular, but the algorithm is clearly not doing the math correctly for some reason.  Not sure what is causing the glitch.  I can go into more detail, but I don't want to bore you. 

Please feel free to contact me if you have any questions, or if it would be helpful for me to provide you with any additional information...  I do have the math worked out and I would be happy to share it with you...

**Reply by erzich** [reply]
Please let me know if there is anything else I can provide that would be helpful to you in your review...

**Reply by erzich** [reply]
Hi - I provided a support ticket on this with openseas, with the ticket number being 751709. I provided you with all the math and the methodology. Can you kindly give me an update or let me know if you are still evaluating this?

Thank you in advance.

**Comment by erzich** [comment]
What do I need to do to get in touch with someone here? Is there someplace I can chat with your team? I’m just looking for some answers to reasonable questions.

**Comment by BeautifoolData** [comment]
Is there a way to exclude traits from the rarity calculations?

**Reply by impreso** [reply]
Exclude trait count you mean?

**Reply by BeautifoolData** [reply]
Some collections have dynamic attributes which causes rarity confusion. MoonCats comes to mind, on-chain accessories can be purchased and displayed. MoonCats can also go from unnamed to named, which is another attribute. However, there are a number of non-changing attributes like Coat Pattern, Face, Rescue Year, etc. Allowing the collection to exclude some attributes from the rarity calculation may be beneficial for users.

**Comment by dazligth** [comment]
Hi there,

in the comment of "open_rarity\scoring\scoring_handler.py" it says:
"Calculates the entropy of the collection, defined to be the sum of the probability of every possible attribute name/value pair that occurs in the collection **times that square root** of such probability."

But where is the "square root" operation done in the code? I can't find it.
Where is the missing piece I don't see? Is it just a typo and "log2​" is meant?
